pandas/tests/base/test_ops.py

```diff
@@ -557,66 +557,25 @@ def test_value_counts_datetime64(self, index_or_series):
         result2 = td2.value_counts()
         tm.assert_series_equal(result2, expected_s)
 
-    def test_factorize(self):
-        for orig in self.objs:
-            o = orig.copy()
-
-            if isinstance(o, Index) and o.is_boolean():
-                exp_arr = np.array([0, 1] + [0] * 8, dtype=np.intp)
-                exp_uniques = o
-                exp_uniques = Index([False, True])
-            else:
-                exp_arr = np.array(range(len(o)), dtype=np.intp)
-                exp_uniques = o
-            codes, uniques = o.factorize()
-
-            tm.assert_numpy_array_equal(codes, exp_arr)
-            if isinstance(o, Series):
-                tm.assert_index_equal(uniques, Index(orig), check_names=False)
-            else:
-                # factorize explicitly resets name
-                tm.assert_index_equal(uniques, exp_uniques, check_names=False)
-
-    def test_factorize_repeated(self):
-        for orig in self.objs:
-            o = orig.copy()
-
-            # don't test boolean
-            if isinstance(o, Index) and o.is_boolean():
-                continue
-
-            # sort by value, and create duplicates
-            if isinstance(o, Series):
-                o = o.sort_values()
-                n = o.iloc[5:].append(o)
-            else:
-                indexer = o.argsort()
-                o = o.take(indexer)
-                n = o[5:].append(o)
+    @pytest.mark.parametrize("sort", [True, False])
+    def test_factorize(self, index_or_series_obj, sort):
+        obj = index_or_series_obj
+        result_codes, result_uniques = obj.factorize(sort=sort)
 
-            exp_arr = np.array(
-                [5, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9], dtype=np.intp
-            )
-            codes, uniques = n.factorize(sort=True)
+        constructor = pd.Index
+        if isinstance(obj, pd.MultiIndex):
+            constructor = pd.MultiIndex.from_tuples
+        expected_uniques = constructor(obj.unique())
 
-            tm.assert_numpy_array_equal(codes, exp_arr)
-            if isinstance(o, Series):
-                tm.assert_index_equal(
-                    uniques, Index(orig).sort_values(), check_names=False
-                )
-            else:
-                tm.assert_index_equal(uniques, o, check_names=False)
+        if sort:
+            expected_uniques = expected_uniques.sort_values()
 
-            exp_arr = np.array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1, 2, 3, 4], np.intp)
-            codes, uniques = n.factorize(sort=False)
-            tm.assert_numpy_array_equal(codes, exp_arr)
+        expected_uniques_list = list(expected_uniques)
+        expected_codes = [expected_uniques_list.index(val) for val in obj]
+        expected_codes = np.asarray(expected_codes, dtype=np.intp)
```

Comment:
Can you not just use np.take here instead?

Reply:
you can use expected_uniques.take is better

Reply:
I don't think I can use `take` here. I want to construct an array containing the indices in `expected_uniques` of the values of `obj`.

```
# given
obj = Series([1, 2, 1, 3, 5])
expected_uniques = obj.unique()  # array([1, 2, 3, 5])

# needed
expected_codes = array([0, 1, 0, 2, 3])
```

I could only use `take` if already have the indices and need the values. I basically need the reverse of `take`.

I guess I could use `where` somehow, but it will probably be more complex than just using vanilla python `list.index()`

Reply:
I think you can just use `pd.factorize` then

```python
In [7]: import pandas as pd
In [8]: obj = pd.Series([1, 2, 1, 3, 5])
In [10]: pd.factorize(obj)
Out[10]: (array([0, 1, 0, 2, 3]), Int64Index([1, 2, 3, 5], dtype='int64'))
```

Reply:
I'm testing factorize here, so I need an alternative implementation 😄

Reply:
ok you can actually use ``.get_loc`` but a simple impl is better; can you add a comment explaining what you are doing (factorizing)